src/librustc_mir/transform/const_prop.rs

```diff
@@ -567,22 +559,15 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
             }
         }
 
-        // If overflow checking is enabled (like in debug mode by default),
-        // then we'll already catch overflow when we evaluate the `Assert` statement
-        // in MIR. However, if overflow checking is disabled, then there won't be any
-        // `Assert` statement and so we have to do additional checking here.
-        if !overflow_check {
```

Comment:
I also just realized this entire `!overflow_check` makes no sense...

It gets initialized via
```rust
let overflow_check = self.tcx.sess.overflow_checks();
```
but whether this MIR has overflow checks is not a global property! In release builds, function bodies have no overflow checks but `const` bodies still have overflow checks.

Ah, but maybe const-prop does not run on `const` bodies? At least with associated const though, one could conceivably still want to optimize the MIR even though it cannot be executed... if we don't that will at the very least mean that `exceeding_bitshifts` does not happen there.

And the other part that makes no sense is causing https://github.com/rust-lang/rust/issues/69020: even with `!overflow_check`, division and remainder still get overflow checks.

And of course, for addition, subtraction and multiplication in MIR with overflow checks, `check_binary_op` will anyway not be called because those will use `CheckedBinOp`, not `BinOp`! So, the check is mostly dead code anyway.

We could keep piling up more and more special cases, but this is getting ridiculous. I feel this requires a fundamentally different approach.

Reply:
So... my gut feeling is that we should stop emitting errors at the `Assert` terminators (which only sometimes exist), and always use the arithmetic operation itself as the source of the error (as that will always exist in one form or another, for debug or release builds). Does that sound sensible? (Yes that is a separate PR.)

Reply:
hmm... That will also solve the situation where we have code paths only run by const prop, because const eval will now also take them. So essentially assert terminators become gotos for miri? Can this work for length checks and such, too?

Reply:
I have no idea what you mean... I was talking only about const-prop not emitting errors at `Assert` terminators, not Miri in general. Miri in general should implement the Rust Abstract Machine, and there arithmetic is wrapping.

Reply:
Ok then: const-prop skips assert terminators and treats them like `Goto`?

Reply:
Yes.

Reply:
> Ok then: const-prop skips assert terminators and treats them like `Goto`?

But doesn't that make "this may panic at runtime" messaging a complete lie?
*Only* `Assert` terminators are the ultimate source of truth for that.

Reply:
I guess there are two categories of `Assert` terminators:
* always emitted (indexing, division by 0, etc.)
* overflow checks, only in debug mode or with a flag enabled etc.

I don't care *too* much about what happens with the latter, but I think the former should still be relied upon.

Reply:
> But doesn't that make "this may panic at runtime" messaging a complete lie?
Only Assert terminators are the ultimate source of truth for that.

Fun story: currently, we do *not* write "this will panic at runtime" when we get our info from `Assert` terminators. We only write that when we do *not* get it from assert terminators.

> I don't care too much about what happens with the latter, but I think the former should still be relied upon.

I can see that. However, then we need some other way to fix https://github.com/rust-lang/rust/issues/61821. Though to be fair that has a different cause... considering that const-prop currently stays *silent* when evaluating a div-by-0 (relying on the panic instead), it probably makes most sense to also make const-prop stay *silent* when evaluating an out-of-bounds access (relying on the panic instead).

Reply:
> Fun story: currently, we do not write "this will panic at runtime" when we get our info from Assert terminators. We only write that when we do not get it from assert terminators.

That's... the wrongest it can be, wow!

> considering that const-prop currently stays silent when evaluating a div-by-0 (relying on the panic instead), it probably makes most sense to also make const-prop stay silent when evaluating an out-of-bounds access (relying on the panic instead).

Yes, all "`Assert` guards UB" situations should be treated the same.

Reply:
On a second thought, this makes no sense... we already don't show out-of-bounds indexing in const-prop. The error there is duplicated because the MIR got duplicated into the promoted. The fix will have to be something promoted-specific.